### PR TITLE
sip 4.17

### PR DESCRIPTION
**Pull request by bfontaine**
This might help fixing https://github.com/Homebrew/homebrew-python/pull/258.

### CODE REVIEW

**Comment by bfontaine** [comment]
```
==> /usr/bin/clang++ -shared -o libtest.dylib test.cpp
==> /usr/local/Cellar/sip/4.17/bin/sip -b test.build -c . test.sip
==> python generate.py
==> make -j1 clean all
rm -f test.so
rm -f siptestcmodule.o
rm -f siptestTest.o
c++ -c -pipe -fPIC -Os -Wall -W -DNDEBUG -I. -I/usr/local/Cellar/sip/4.17/include -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -o siptestcmodule.o siptestcmodule.cpp
c++ -c -pipe -fPIC -Os -Wall -W -DNDEBUG -I. -I/usr/local/Cellar/sip/4.17/include -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -o siptestTest.o siptestTest.cpp
c++ -headerpad_max_install_names -bundle -undefined dynamic_lookup -o test.so siptestcmodule.o siptestTest.o -L. -ltest
==> python run.py
Traceback (most recent call last):
  File "run.py", line 1, in <module>
    from test import Test
ImportError: dlopen(/private/tmp/sip20151125-97261-1jrba4w/test.so, 2): Library not loaded: libtest.dylib
  Referenced from: /private/tmp/sip20151125-97261-1jrba4w/test.so
  Reason: unsafe use of relative rpath libtest.dylib in /private/tmp/sip20151125-97261-1jrba4w/test.so with restricted binary
Error: sip: failed
```

**Comment by bfontaine** [comment]
Adding `--static` to `configure.py` fixes the build but breaks the tests :disappointed:

**Comment by bfontaine** [comment]
cc @tdsmith, the local Python expert :smile:

File: Library/Formula/sip.rb [hunk 1]
```diff
@@ -13,17 +14,15 @@ class Sip < Formula
...
   depends_on :python => :recommended if MacOS.version <= :snow_leopard
   depends_on :python3 => :optional
 
-  if build.without?("python3") && build.without?("python")
-    odie "sip: --with-python3 must be specified when using --without-python"
-  end
-
   def install
+    if build.without?("python3") && build.without?("python")
+      raise "sip: --with-python3 must be specified when using --without-python"
```
Comment (UniqMartin) [inline Library/Formula/sip.rb]: It seems that with Mike's latest documentation overhaul we have settled on `odie` for these kinds of messages.

**Comment by UniqMartin** [comment]
Ironically, the test has been shot down by SIP (system integrity protection). Here's what you want to change to make it work:

```diff
diff --git i/Library/Formula/sip.rb w/Library/Formula/sip.rb
index 591c5bfb..737c3367 100644
--- i/Library/Formula/sip.rb
+++ w/Library/Formula/sip.rb
@@ -94,7 +94,8 @@ class Sip < Formula
       t = Test()
       t.test()
     EOS
-    system ENV.cxx, "-shared", "-o", "libtest.dylib", "test.cpp"
+    system ENV.cxx, "-shared", "-Wl,-install_name,#{testpath}/libtest.dylib",
+                    "-o", "libtest.dylib", "test.cpp"
     system "#{bin}/sip", "-b", "test.build", "-c", ".", "test.sip"
     Language::Python.each_python(build) do |python, version|
       ENV["PYTHONPATH"] = lib/"python#{version}/site-packages"
```

(This adds an absolute install name to the dylib. Its absence is what the dynamic linker is complaining about when testing with the SIP-protected system Python.)

File: Library/Formula/sip.rb (hunk 1 above)
Comment (bfontaine) [inline Library/Formula/sip.rb]: No, Xu advised to use `raise` because it’s easier to manage: we can catch exceptions and provide friendly messages.

**Comment by bfontaine** [comment]
Thank you @UniqMartin! :smile:

File: Library/Formula/sip.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Formula/sip.rb]: I know, see https://github.com/Homebrew/homebrew/pull/44954#discussion_r42487584 and the ensuing discussion.

Comment (bfontaine) [inline Library/Formula/sip.rb]: Thanks; I reverted back to `odie`.

**Pull request closed without merging**